src/util/jsonLog.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+
+import stringify from "json-stable-stringify";
+import fs from "fs-extra";
+import * as C from "./combo";
+
+/**
+ * JsonLog tracks and serializes append-only logs of JSON objects.
+ *
+ * At its heart, it's basically a simple wrapper around an array, which
+ * enforces the rule that items may be appended to it, but never removed.
+ *
+ * It also provides serialization logic. While the log is an array of JSON
+ * objects and thus a JSON object itself, we don't want to store it on disk as
+ * a stringified JSON object. That would lose the structure of an append-only
+ * array, as every change would result in a rewrite of a single massive line of
+ * text, and would be inefficient in Git storage, and make for ugly diffs and
+ * merge conflicts.
+ *
+ * Instead, we store each item on its own line in the output format, using
+ * `stringify` on the individual JSON objects. We support // comment syntax,
+ * and the toString and writeJsonLog methods allow you to provide prefix
+ * comments; we recommend putting prefix comments like
+ * ["AUTOGENERATED FILE", "do not manually edit"].
+ *
+ * The JsonLog is always parsed using a Combo.Parser, which ensures type safety
+ * at runtime.
+ */
+export class JsonLog<T: C.JsonObject> {
```

Comment:
`<T: C.JsonObject>` has the limitation that events can't have a custom `toJSON` such as for a `Set`, while they will require a custom parser per event type. Have a hunch we can group together serializing and parsing per event in a comprehensive way.

Reply:
I've talked with @wchargin about making a dual to Combo.Parser that would bundle serialization in too. Which I'd really like. He says it's not trivial but should be possible, but I didn't follow all the details.

If we need to support objects like sets, we could make a version of JsonLog where we don't assume T is a C.JsonObject, but in the `toString` method, you need to provide a serializer that maps `T => JsonObject`. Should be easy to add if/when we actually need it.